### ISSUE TITLE
__add__

### ISSUE THREAD

**Issue by Lex-2008** [uliss4ever/My_repository_2#2]
https://github.com/uliss4ever/My_repository_2/blob/d37f4f70be88848a58949f2d462da777e9356474/class_Date.py#L150-L153

> d2 += TimeDelta), и вот у нас есть новый инстанс, не?

Неет, помнишь что мы про эти переменные говорили, что после `d1 = d2` они ссылаются на один и тот же объект в памяти и изменение одной переменной будет видно в другой? Вот этот пример:

```
d1 = Date(1, 1, 2001)
d2 = d1       # d1 и d2 - это одна и та же дата:
d2.day += 2   # если поменять день в одной,
              # то это будет видно и в другой.
print(d1)     # выведет "03.01.2001"
```

> Или если потом мы поменяем d1, то и d2 за ней поменяется?

Да, вот именно так и будет.

Так, ну, во-первых, функция `__add__` у нас будет состоять из двух частей: `c = a` и `c += b`.

Давай сначала подумаем о строчке `c = a`. Здесь нам надо **создать новый объект** класса Date, равный `a`. Как мы создаём новый объект класса Date?

**Comment by uliss4ever** [comment]
d3 = Date(day, month, year) - вот так?
А d1=d2 - это я имела в виду, а=с

**Comment by Lex-2008** [comment]
Точно!

Так, `d3` мы создали, а как теперь сделать его точной копией `self`? Установить ему day, month, year, равные day, month, year, записанные в `self`.

Как мы можем устанавливать day, month, и year у `d3`?

И как мы можем узнать, какие day, month, и year записаны в `self`?

**Comment by uliss4ever** [comment]
Я об этом думала, надо точно такие же.
d3 = Date(self.day, self.month, self.year)
Или это глупость?

**Comment by uliss4ever** [comment]
Но просто у d1 есть self и у d3 есть self. Как сделать чтобы у d3 был self от d1 ?
И как же это охренительно интересно...

**Comment by Lex-2008** [comment]
> d3 = Date(self.day, self.month, self.year)

Да, именно так!

**Comment by Lex-2008** [comment]
>  у d1 есть self и у d3 есть self.

Не совсем так. "self" - это указатель на тот объект, который стоит перед точкой при вызове метода. Например, у нас есть класс "корова", у неё есть метод "махнуть хвостом", и два объекта этого класса: корова Бурёнка и корова Машка.

Вот если мы теперь напишем:
```
Бурёнка.махнуть_хвостом()
```

то в методе "махнуть хвостом" `self` будет указывать на корову Бурёнку; `self.хвост` будет указывать на хвост коровы Бурёнки, и `self.хвост.повернуть(90°)` повернёт хвост коровы Бурёнки на 90°.

Если же написать:
```
Машка.махнуть_хвостом()
```

то будет всё так же, только повернётся хвост коровы Машки.

Если же у коровы будет метод "махнуть чужим хвостом", то ему нужно будет передать дополнительный параметр - корову, чьим хвостом надо махнуть.

**Comment by uliss4ever** [comment]
(валяюсь от смеха...)
Вот что тогда, мне надо чтобы d3 махнуло хвостом от d1, так же?

**Comment by uliss4ever** [comment]
d3 = Date(self.day(d1), self.month(d1), self.year(d1))  приходит в голову. Но это не то, day же это не метод? Это аргумент? Прости, если я порю чушь

**Comment by Lex-2008** [comment]
(а я испугался - куда ты пропала…)

Да, надо чтобы они были независимыми. И правильный ответ ты уже написала:

> d3 = Date(self.day, self.month, self.year)

Больше думать не надо, остановись! :laughing:

**Comment by Lex-2008** [comment]
Давай вспомним: функция `__add__` должна сделать `c = a + b`, то есть создать новый объект (назовём его `c` или `d3`, не важно), и при этом не изменить объекты `a` и `b`.

Для этого мы сначала делаем `c = a`, то есть создаём новый объект с day, month, year, взятыми из `a` (который self). Ты уже написала, как.

А потом прибавляем к нему `b`, который TimeDelta - для этого мы будем использовать функцию, которую напишем потом, в пятницу. Это будет просто `d3 += other`.

**Comment by uliss4ever** [comment]
У меня тут просто процесс самопознания параллельно. Обострился. Я рыдала. Не получился генератор. И не прописываются эти пути к скрипту.
Лёш, ты суперски объясняешь)

**Comment by Lex-2008** [comment]
Господи, какой генератор? Ты сколько заданий одновременно делаешь и почему они у тебя в голове не перемешиваются?

> не прописываются эти пути к скрипту.

это про тесты? Удалить их из раздела "Python" не помогает? https://github.com/uliss4ever/My_repository_2/issues/1#issuecomment-786196854

**Comment by uliss4ever** [comment]
Нет, не помогает. Генератор случайных адресов - это по позапрошлому 110 курсу, хвост. Это я тебе позже хотела показать. И даже кое-что спросить. Он получился, но не совсем. Иногда мне кажется, я не доживу до конца этого курса, честно. Но я обещала Любе (это которая учила нас два первых курса) что не сдамся, держу слово) 
Ты ещё здесь или уже пора сворачиваться?
Зато я там написала фабрику декораторов, на этот недогенератор

**Comment by Lex-2008** [comment]
Точно не помогает? Пришли скриншот. Может быть, надо закрыть-открыть Пичарм?

Да, наверное, скоро позовут сворачиваться - если вдруг перестану отвечать - значит, свернули :)

фабрика декораторов - звучит круто, если это то, о чем я думаю.

**Comment by uliss4ever** [comment]
Это то, о чём ты думаешь и он получился, потом покажу. Офигительная вещь, я прямо без ума, что можно такое сочинить (это я не про себя, а про того, кто писал Питон)

**Comment by uliss4ever** [comment]
Да я пока не поняла, где тут вешать скрины. Разберусь пришлю, но там Енот сказал, нужен файл setup (хотя нам показывал без него и всё работало). Пасиб тебе.

**Comment by uliss4ever** [comment]
> (а я испугался - куда ты пропала…)
> 
> Да, надо чтобы они были независимыми. И правильный ответ ты уже написала:
> 
> > d3 = Date(self.day, self.month, self.year)
> 
> Больше думать не надо, остановись! 😆

Я тут чего не понимаю, хоть и вроде правильное написала. А где видно, что day - он именно от d1 ? А если у нас два объекта созданы, d1 и d2. Мы их оба создаём с помощью day, month, year. А потом d3. С теми же. Какая-то отличительная черта нужна, по-моему. То есть вот в этой строке  d3 = Date(self.day, self.month, self.year) это чей у нас self - d1 ? А почему не d2 , если у нас два объекта? И где у этого d3 тогда свой self...

Это вот, ниже, на всякий случай вынесла, мне про это подумать надо)
d1 = Date(30, 12, 2021)
d1.day = 31
d2 = Date(1, 2, 2020)
d2.day = 29

* Кстати, вот думаю, ещё в 14 веке отцы-богословы (ну, один, Оккам) советовали не творить лишних сущностей. Интересно, про Питон может они тоже догадывались, что тут это важно?

**Comment by Lex-2008** [comment]
> d3 = Date(self.day, self.month, self.year) это чей у нас self - d1 ? А почему не d2 , если у нас два объекта?

Да, вспомни пример с коровой:

```
Class корова:
  def махнуть_хвостом(self):
    self.хвост.повернуть(90°)
```

self - это какая корова?

На русский язык этот код переводится так: "если попросить корову махнуть хвостом, _она_ повернет хвост на 90°". Кто "она"? Какая корова повернёт хвостом?

Правильный ответ: "та, которую попросили". Если попросить Бурёнку - махнёт Бурёнка; если попросить Машку - махнёт Машка. Внутри метода "махнуть_хвостом" корова, которую попросили махнуть хвостом, доступна под именем "self".

Может быть, понятней будет, если "self" перевести на русский как "я"? У `d1` есть своё "я", у `d2` есть своё "я", у `d3` и любой даты (или коровы), которую мы создадим, есть своё "я".

Давай ещё раз вернёмся к твоему вопросу:

> d3 = Date(self.day, self.month, self.year) это чей у нас self - d1 ? А почему не d2 , если у нас два объекта?

Мы эту строчку пишем внутри функции `def __add__(self, other: TimeDelta) -> "Date":`. Здесь self - это тот объект (та дата), которая стоит слева, когда мы вызвали этот метод. Как в примере с коровой: `Бурёнка.махнуть_хвостом()` - значит, `self` - это Бурёнка. Если бы мы написали `Машка.махнуть_хвостом()` - то `self` был бы Машкой. Так же и тут - только `__add__` вызывается не через точку, а плюсом:

`d1 + что_то_там` - `self` будет d1, `d2 + что_то_там` - `self` будет d2. Просто `+` висящий в воздухе написать нельзя - нужно обязательно что-то перед ним поставить (ну и после него тоже) - и то, что указано перед плюсом и будет `self`.

**Comment by uliss4ever** [comment]
Да, поняла) Про self я кстати вообще понимаю, что это, ну типа «сам», «тот», именно тот, с которым мы сейчас взаимодействуем, а не какой-то другой. «Я» объекта, в общем, лучше не скажешь. Просто когда ты сказал про дополнительный параметр, вот тут я загрузилась. Потому что я пока ещё в тех, своих, «до-ооп-шных» категориях существую. Но уже намного легче стало, переключаюсь.

**Comment by Lex-2008** [comment]
Да, не знаю, насколько это полезно, но я просто пытаюсь тебе показать, что ооп - это просто дополнительные слова и ограничения нанесенные поверх обычных "до-ооп-шных" функций. Например, если посмотреть на твою [очередь с приоритетами][pq] - там во всех функциях используется глобальный список priority_queue.

[pq]: https://github.com/uliss4ever/111-Py/blob/master/Tasks/a2_priority_queue.py#L9-L10

А что делать, если нам в какой-то программе надо использовать две такие очереди?

Либо сделать копию каждой функции, которая использовала бы другую глобальную переменную,

либо каждой функции добавить ещё один аргумент - список priority_queue, который надо использовать. Вот так:

```
def enqueue(priority_queue, elem: Any, priority: int = 0) -> None:
  …
def dequeue(priority_queue) -> Any:
  …
def peek(priority_queue, ind: int = 0, priority: int = 0) -> Any:
  …
def clear(priority_queue) -> None:
  …
```

Причем внутренности каждой функции меняться не будут - просто "priority_queue" внутри каждой функции будет означать "первый агрумент, переданный в функцию" вместо "глобальная переменная `priority_queue`".

Пока понятно? Точно понятно? :) Или лучше пока этим не грузить?

**Comment by uliss4ever** [comment]
Да, мне это понятно и "грузить", конечно! (только если тебе самому не утомительно, потому что я не всегда с первого раза догоняю, но перечитать 25 раз помогает однозначно). Я очень благодарна. А как мы передадим, так же как объект день, priority_queue1 (или 2 или 3..) да? Я бы очень хотела это ещё на практике попробовать. То есть и попробую, конечно, просто мне пока код не вообразить, где это может быть полезно.
Про эти ограничения - я ещё пока очень синтаксиса боюсь. Стало столько всего, где можно забыть, накосячить, точки, скобки, и интерпретатор тебя поймёт не так и вы не можете договориться...

**Comment by Lex-2008** [comment]
По поводу синтаксиса вспомнился рассказ "телефон" К.И. Чуковского (на самом деле, это первая глава из его книги "Серебряный герб", но я книгу не читал, а "телефон" запомнился: https://librebook.me/serebrianyi_gerb/vol1/1.

Не знаю, что здесь можно посоветовать, кроме очевидных вещей типа того, что надо понимать каждое слово в каждой строчке: это объект, это его свойство, это его метод, и т.д.

> просто мне пока код не вообразить, где это может быть полезно.

В смысле, кому нужны очереди с приоритетами? Ну, вообще, "очередь" в программировании используется тогда же, как и в магазине: когда есть один обработчик (кассир), к которому в очередь выстраиваются задания (покупатели с корзинами). Задания-покупатели ждут, когда обработчик-кассир обработает одно(го) и перейдёт к следующему.

Ну а приоритет появляется, когда одно задание считается более важным, чем другие. Например, когда пользователь сидит за компьютером, запускает ворд, а у него при этом в фоне музыка играет и архиватор что-нибудь сжимает. И вот жесткому диску одновременно приходят три запроса на чтение файлов: архиватор хочет заархивировать следующий файл; музыкальный плеер переходит к следующей части музыкального файла, и ещё ворд хочет что-то показать (либо свой интерфейс, либо файл который он открывает). Жесткий диск должен правильно расставить приоритеты: первым идёт плеер, чтобы музыка не дёргалась; на втором месте ворд, т.к. на него смотрит пользователь; и на последнем месте арзиватор, т.к. ему все равно ещё долго работать. Эти приоритеты обычно задают сами программы, а система (Windows/MacOS) их немного правит - немного повышает приоритет тому приложению, с которым сейчас работает пользователь, чтобы оно было чуть более "отзывчивым".

**Comment by uliss4ever** [comment]
Ну да, то есть я про очередь понимаю, но я же сейчас не напишу такое ... (ну что там за работу операционки отвечает). А так я в целом пока писала, врубилась. Очень сложно было расставить эти очереди внутри одной очереди и добиться, чтобы там всё адекватно добавлялось и удалялось. Особенно долго реек не получался. То есть можно было реализовать проще, несколько очередей сделать, фором, для каждого приоритета своя. Но я хотела именно ту реализацию, которая у меня в итоге есть)
Рассказ пойду почитаю, пасиб)
... огонь просто. Чуковский неподражаем)

**Comment by Lex-2008** [comment]
Да, несколько очередей - это неспортивно, у тебя гораздо круче получилось!

Ну так вот по поводу синтаксиса - вот это у ребят в рассказе, кто запятые в середину слов вставлял, проблемы с синтаксисом :) А у тебя норм.

Да, я правда видел людей, которым надо объяснять, что если в строке количество открывающихся скобок не равно количеству закрывающихся - то это ошибка. `for(x in a` - по-моему, это просто некрасиво. https://xkcd.com/859/

Так что не переживай: со временем насмотришься на много кода и на интуитивном уровне начнешь синтаксис видеть: что "красиво", а что нет :)

Я, правда, ООП начал смотреть лет через 5~7 после того, как программированием начал заниматься, а у тебя всё гораздо быстрее идёт!

**Comment by uliss4ever** [comment]
Да ты смеёшься...:) никуда оно не идёт, меня по нему за уши волокут со страшной скоростью, невзирая на мои вопли «остановитесь, погодите, дайте на сущности посмотреть!». Вот доволокли до того, что уже завтра веб. Ну, то есть жаловаться нельзя, я сама этого хотела. Но, Лёш. Ничего у меня вчера всё-таки не вышло. То есть инит, я сломала, ну мне так кажется. В 36 строчке ругается. Но дальше без идей и без сил. И тесты не могу написать, потому что как тестировать, если код ещё не доделала. То есть нам объясняли, что так можно. Наоборот, сперва тесты, а потом код. Но это не я пока, кто так может. Поэтому если ты не против, давай постепенно и с тобой разбираться. Не успею к 6-му - значит, когда успеем. Делала ещё пакет для Трависа, всё вроде повторила за Енотом, но жёлтой точечки, как он обещал, не появилось. Сегодня буду драйвера писать (о Боже, это точно я?))). И генератор, там я хоть понимаю, что делать...

**Comment by uliss4ever** [comment]
Да, а про скобки, ты видел там у нас даже специальное задание есть? Тоже в прошлом курсе. Я его ещё с помощью очереди (простой) решила, но Еноту понравилось больше это, сказал его оставить. 
Вообще я везде постоянно вижу, что то, что правильно и логично, то обязательно ещё очень и очень красиво. Питон очень красивый. И вообще всё программирование. Когда я успеваю остановиться и подумать, то думаю, ну как они такое изобрели? Вон, искусственный язык не смогли толком придумать, эсперанто есть, но это скорее какая-то упрощенная компиляция из того, что уже было. А тут - целые системы, сложные, продуманные и работающие как надо. И при этом - совершенно с нуля, на ровном месте, можно сказать. В том смысле, что у человечества вообще ничего подобного ещё 100 назад не было. Только логика. Короче, космос)

**Comment by Lex-2008** [comment]
Да, про скобки задание видел, с помощью очереди (а точнее стека - стопки) его решать действительно "слишком круто", но стек пригодится, если надо проверять строки с разными типами скобок, например `({)}` - тут _количество_ скобок правильное, но вот последовательность не та.
